src/boot/efi/devicetree.c

```diff
@@ -0,0 +1,173 @@
+        EFI_CONFIGURATION_TABLE *entry = ST->ConfigurationTable;
+        EFI_CONFIGURATION_TABLE *end = entry + ST->NumberOfTableEntries;
+
+        for (; entry < end; entry++) {
+                if (CompareGuid(&entry->VendorGuid, &(EFI_GUID)DEVICE_TREE_GUID) == 0) {
+                        state->orig = entry->VendorTable;
+                        return EFI_SUCCESS;
+                }
+        }
+
+        return EFI_UNSUPPORTED;
+}
+
+static EFI_STATUS devicetree_allocate(struct devicetree_state *state, UINTN size) {
+        UINTN pages = DIV_ROUND_UP(size, EFI_PAGE_SIZE);
+        EFI_STATUS err;
+
+        err = uefi_call_wrapper(BS->AllocatePages, 4, AllocateAnyPages,
+                        EfiACPIReclaimMemory, pages, &state->addr);
+        if (EFI_ERROR(err))
+                return err;
+
+        state->pages = pages;
+        return err;
+}
+
+static UINTN devicetree_allocated(struct devicetree_state *state) {
+        return state->pages * EFI_PAGE_SIZE;
+}
+
+static VOID *devicetree_ptr(struct devicetree_state *state) {
+        assert_cc(sizeof(VOID *) <= sizeof(state->addr));
+        return (VOID *)(uintptr_t)state->addr;
+}
+
+static EFI_STATUS devicetree_fixup(struct devicetree_state *state, UINTN len) {
+        EFI_DT_FIXUP_PROTOCOL *fixup;
+        UINTN size;
+        EFI_STATUS err;
+
+        err = LibLocateProtocol(&(EFI_GUID)EFI_DT_FIXUP_PROTOCOL_GUID, (VOID **)&fixup);
+        if (EFI_ERROR(err)) {
+                Print(L"Could not locate device tree fixup protocol, skipping.\n");
+                return EFI_SUCCESS;
+        }
+
+        size = devicetree_allocated(state);
+        err = uefi_call_wrapper(fixup->Fixup, 4, fixup,
+                        devicetree_ptr(state), &size,
+                        EFI_DT_APPLY_FIXUPS | EFI_DT_RESERVE_MEMORY);
+        if (err == EFI_BUFFER_TOO_SMALL) {
+                EFI_PHYSICAL_ADDRESS oldaddr = state->addr;
+                UINTN oldpages = state->pages;
+
+                err = devicetree_allocate(state, size);
+                if (EFI_ERROR(err))
+                        return err;
+
+                CopyMem(devicetree_ptr(state), (VOID *)(uintptr_t)oldaddr, len);
+                err = uefi_call_wrapper(BS->FreePages, 2, oldaddr, oldpages);
+                if (EFI_ERROR(err))
+                        return err;
+
+                size = devicetree_allocated(state);
+                err = uefi_call_wrapper(fixup->Fixup, 4, fixup,
+                                devicetree_ptr(state), &size,
+                                EFI_DT_APPLY_FIXUPS | EFI_DT_RESERVE_MEMORY);
+        }
+
+        return err;
+}
+
+EFI_STATUS devicetree_install(struct devicetree_state *state,
+                EFI_FILE_HANDLE root_dir, CHAR16 *name) {
+        _cleanup_(FileHandleClosep) EFI_FILE_HANDLE handle = NULL;
+        _cleanup_freepool_ EFI_FILE_INFO *info = NULL;
+        UINTN len;
+        EFI_STATUS err;
+
+        assert(state);
```

Comment:
(You're still not asserting on name here and the other functions should assert their input too)